From b901ab96a502f4fdc05453604c4cfe545f45d1d9 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Tue, 7 Oct 2025 00:42:07 +0200
Subject: [PATCH 02/17] winex11: Use Xshape fallback on Mutter

Workarounds https://github.com/wine-mirror/wine/commit/35193586, without this
osu! force-enables vsync due to the compositor not being disabled, even in fullscreen.
---
 dlls/winex11.drv/window.c | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 9471971b510..117a7648ffc 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -2618,7 +2618,9 @@ Window get_dummy_parent(void)
         attrib.colormap = default_colormap;
 
 #ifdef HAVE_LIBXSHAPE
+        if (!X11DRV_HasWindowManager( "Mutter" ))
         {
+            TRACE("Not using Xshape fallback, WM isn't Mutter.\n");
             static XRectangle empty_rect;
             dummy_parent = XCreateWindow( gdi_display, root_window, 0, 0, 1, 1, 0,
                                           default_visual.depth, InputOutput, default_visual.visual,
@@ -2628,11 +2630,15 @@ Window get_dummy_parent(void)
             XShapeCombineRectangles( gdi_display, dummy_parent, ShapeInput, 0, 0, &empty_rect, 1,
                                      ShapeSet, YXBanded );
         }
-#else
-        dummy_parent = XCreateWindow( gdi_display, root_window, -1, -1, 1, 1, 0, default_visual.depth,
-                                      InputOutput, default_visual.visual,
-                                      CWColormap | CWBorderPixel | CWOverrideRedirect, &attrib );
-        WARN("Xshape support is not compiled in. Applications under XWayland may have poor performance.\n");
+        else
+        {
+#endif
+            dummy_parent = XCreateWindow( gdi_display, root_window, -1, -1, 1, 1, 0, default_visual.depth,
+                                          InputOutput, default_visual.visual,
+                                          CWColormap | CWBorderPixel | CWOverrideRedirect, &attrib );
+            FIXME("Using Xshape fallback. Either your WM is Mutter, or Xshape support was not compiled in.\n");
+#ifdef HAVE_LIBXSHAPE
+        }
 #endif
         XChangeProperty( gdi_display, dummy_parent, x11drv_atom(_NET_WM_WINDOW_OPACITY),
                          XA_CARDINAL, 32, PropModeReplace, (unsigned char *)&opacity, 1 );
-- 
2.55.0

